### PR TITLE
Add data splitting and train models

### PR DESCRIPTION
**Pull request by jenna-tomkinson**
# Add data splitting and train models

In this PR, the regression model module is added where the data is split and models are trained per nuclear speckle stain feature.

### CODE REVIEW

File: 1.regression_modelling/regression_model_env.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+- conda-forge::nb_conda_kernels
+- conda-forge::scipy
+- conda-forge::numpy
+- conda-forge::scikit-learn
+- conda-forge::seaborn
+- conda-forge::umap-learn
+- conda-forge::matplotlib
+- conda-forge::pyarrow
+- pip:
+  - pycytominer
```
Comment (MikeLippincott) [inline 1.regression_modelling/regression_model_env.yml:22]: consider adding a version here? Might help with reproducibility.

File: 1.regression_modelling/scripts/0.split_data.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+# path to big drive where data is located (large)
+drive_path = pathlib.Path("../../../../media/18tbdrive/")
+
+# set path to folder in image profiling repo with the normalized profiles
+norm_path = pathlib.Path(
+    drive_path / "Github_Repositories/nuclear_speckle_image_profiling/4.preprocess_features/data/single_cell_profiles"
+).resolve(strict=True)
+
+# load all normalized parquet files
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/0.split_data.py:36]: Is each file a plate?

File: 1.regression_modelling/scripts/0.split_data.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+# concatenate them into one data frame
+combined_df = pd.concat(dfs, ignore_index=True)
+
+# perform feature selection to drop any columns that have NaN (avoid downstream issues)
+combined_df = feature_select(
+    combined_df,
+    operation="drop_na_columns",
+    na_cutoff=0
+)
+
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/0.split_data.py:49]: Feature selection maybe belongs in another notebook used during processing?

File: 1.regression_modelling/scripts/0.split_data.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+# In[3]:
+
+
+# load in filtered tuple dataframe
+filtering_df = pd.read_parquet(
+    pathlib.Path(
+        "../0.data_analysis_and_processing/filtered_single_cells/filtered_single_cell_profiles.parquet"
+    )
+)
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/0.split_data.py:65]: What are the filtered sc? Are these QC'd? How is this df different from the feature selected one?

A markdown cell at the top of the notebook explaining the "sub-sections" of this notebook might help

File: 1.regression_modelling/scripts/0.split_data.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+# define the columns to match that are important for filtering
+columns_to_match = [
+    'Metadata_Plate', 'Metadata_Well', 'Metadata_Site',
+    'Metadata_Nuclei_Location_Center_X', 'Metadata_Nuclei_Location_Center_Y'
+]
+
+# ensure both dataframes have these columns
+combined_df_filtered = combined_df[columns_to_match]
+filtering_df_filtered = filtering_df[columns_to_match]
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/0.split_data.py]: This was somewhat confusing for me. Definitely semantics here and naming is hard. `filtering_df_filtered` is difficult for me to understand

File: 1.regression_modelling/scripts/0.split_data.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+# holdout all cells from the 293T cell line as CSV
+holdout_data = filtered_combined_df[filtered_combined_df['Metadata_CellLine'] == '293T']
+
+# Save the holdout data to a CSV file
+holdout_data.to_csv(f"{data_dir}/holdout_data_293T.csv", index=False)
+
+# Print the shape of holdout_data to verify
+print(holdout_data.shape)
+
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/0.split_data.py]: Are there multiple cell types? Is there a rationale for using `.csv` files? If yes, great! If no, consider parquet.

File: 1.regression_modelling/scripts/0.split_data.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+remaining_data = filtered_combined_df[filtered_combined_df['Metadata_CellLine'] != '293T']
+
+# Split the data into training and testing sets
+train_data, test_data = train_test_split(
+    remaining_data,
+    test_size=0.3,
+    random_state=0,  # For reproducibility
+    shuffle=True     # Ensure data is shuffled before splitting
+)
+
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/0.split_data.py:134]: While doing your data splits, are there any variables that the data should be stratified over? 

If there are multiple cell types it might be worth ensuring that each cell type has the same ratio of representation in each split. Same for treatment (drug or RNAi or CRISPRi or ORF, etc) 

You know the data best, so I will leave this up to you to decide.

File: 1.regression_modelling/scripts/0.split_data.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+# Save the training and testing data to CSV files
+train_data.to_csv(f"{data_dir}/training_data.csv", index=False)
+test_data.to_csv(f"{data_dir}/testing_data.csv", index=False)
+
+# Print the shapes of the splits to verify
+print(f"Training data shape: {train_data.shape}")
+print(f"Testing data shape: {test_data.shape}")
+
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/0.split_data.py]: same here with the `.csv`

File: 1.regression_modelling/scripts/1.train_models.py [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+
+import pandas as pd
+import pathlib
+import numpy as np
+import joblib
+from sklearn.linear_model import ElasticNet
+from sklearn.model_selection import KFold, RandomizedSearchCV
+import warnings
+import sys
+import os
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/1.train_models.py]: consider using isort here

File: 1.regression_modelling/scripts/1.train_models.py [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+# Ignore the ConvergenceWarnings (only thing that will work 🙃)
+if not sys.warnoptions:
+    warnings.simplefilter("ignore")
+    os.environ["PYTHONWARNINGS"] = "ignore"
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/1.train_models.py:30]: Looks good!

File: 1.regression_modelling/scripts/1.train_models.py [hunk 11]
```diff
@@ -0,0 +1,213 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# # Train ElasticNet model to predict each A647 and GOLD feature using the nuclear features
+# 
+# In this notebook, we split the features into each group for nuclear speckle or nucleus and then train a model per nuclear speckle feature using the nuclear features to predict it.
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/1.train_models.py]: Might be worth specifying the model type here if it is a classifier or a regression model.

File: 1.regression_modelling/scripts/1.train_models.py [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+    os.environ["PYTHONWARNINGS"] = "ignore"
+
+
+# ## Set paths and random seed
+
+# In[2]:
+
+
+# set numpy seed to make sure any random operations performs are reproducible
+np.random.seed(0)
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/1.train_models.py:39]: Whoop!

File: 1.regression_modelling/scripts/1.train_models.py [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+
+# In[5]:
+
+
+# Suppress all warnings
+warnings.filterwarnings("ignore")
+
+for a647_feature in a647_features:
+    y = training_df[a647_feature]
+
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/1.train_models.py]: I love this approach, we did this for the Pyroptosis paper. I also think in a future PR it might be worth exploring a multi-regression model where Nuclei features predict all the A647 features in one model. You would probably take a hit on performance in fovour of generalizability. This will be up to you if it is worth a future PR. Might be worth it and fit well in a supplementary figure! XD

File: 1.regression_modelling/scripts/1.train_models.py [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+# ## Train A647 models
+
+# In[5]:
+
+
+# Suppress all warnings
+warnings.filterwarnings("ignore")
+
+for a647_feature in a647_features:
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/1.train_models.py]: personal preference but could add a tqdm bar here to monitor progress in the list of models being trained!

File: 1.regression_modelling/scripts/1.train_models.py [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+    shuffled_model_filename = (
+        shuffled_dir / f"{gold_feature}_shuffled_tuned_model.joblib"
+    )
+    joblib.dump(random_search_shuffled.best_estimator_, shuffled_model_filename)
+
+    # Print confirmation
+    print(f"Trained and saved tuned model for {gold_feature}")
+    print(f"Trained and saved shuffled tuned model for {gold_feature}")
+
+print("All GOLD models have been trained and tuned!")
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/1.train_models.py]: Consider adding the gold and A647 feature lists together in one list since they both use the same logreg params.
This would reduce code and DNR,

**Review by MikeLippincott — CHANGES_REQUESTED** [review]
Looks really good and I am excited to see the model evaluations! I am requesting changes mainly for the most important step - the data splitting. 

I do not have enough information about the Metadata to know if this is the right way to split the data. So if this is the best way, my apologies! Please ping me if you have any questions, we can always talk tomorrow more.

File: 1.regression_modelling/regression_model_env.yml (hunk 1 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/regression_model_env.yml:22]: This is a bit of a slippery slope for me. Greg defaults more to leaving environments open, which I see, while others in the lab prefer to have more structured versioning. 

I think this repo is in its early stages and I think that this is beyond the scope of this PR. I am happy to turn this into an issue for down the road. Thanks!

File: 1.regression_modelling/scripts/0.split_data.py (hunk 2 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/0.split_data.py:36]: Yes, each file relates to a "slide" or plate in our terminology.

File: 1.regression_modelling/scripts/0.split_data.py (hunk 3 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/0.split_data.py:49]: Since we are using normalized data to generate these models, there is a likelihood that there are NaNs in the data.

Instead of using a different method, I use pycytominer to clear out columns with NaNs prior to training. So it would probably best fit in this notebook since it is a small step in the process.

File: 1.regression_modelling/scripts/0.split_data.py (hunk 4 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/0.split_data.py:65]: I can add a quick snippet of information at the top of the file! Thanks for catching that I didn't have that!

To explain here, for the methodology that Cameron will use with the images, he needs to clear out more single-cells that are out of a specific range of width and height for the crop. This is what is known as the "filtering_df". 

To make sure that both of the methods (mine and Cameron's) is comparable, we have to use the same single-cells which is why we have a filtering dataframe to filter down the morphological profiles.

File: 1.regression_modelling/scripts/0.split_data.py (hunk 5 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/0.split_data.py]: Fair point! These variables are pretty unnecessary, so I have removed them to avoid more confusion. Thanks for pointing out!

File: 1.regression_modelling/scripts/0.split_data.py (hunk 6 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/0.split_data.py]: There are two cell types, and we are holding out one of them.

No rationale for using CSV, that is a super easy change!

File: 1.regression_modelling/scripts/0.split_data.py (hunk 7 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/0.split_data.py:134]: Fair points!

There are only siRNA-treated cells left after removing holdout from one cell line. We don't have any real reason to stratify these treatments in a specific way, so for now this is pretty good enough. Thank you for bringing up these points!

File: 1.regression_modelling/scripts/0.split_data.py (hunk 8 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/0.split_data.py]: Updated!

File: 1.regression_modelling/scripts/1.train_models.py (hunk 9 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/1.train_models.py]: Will do!

File: 1.regression_modelling/scripts/1.train_models.py (hunk 10 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/1.train_models.py:30]: Took me forever to get that to work haha, so thank you!

File: 1.regression_modelling/scripts/1.train_models.py (hunk 11 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/1.train_models.py]: Added!

File: 1.regression_modelling/scripts/1.train_models.py (hunk 12 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/1.train_models.py:39]: 🎉

File: 1.regression_modelling/scripts/1.train_models.py (hunk 13 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/1.train_models.py]: I so agree! It would be really cool to explore so I think this will be kept in mind for a future paper, thank you lots!

File: 1.regression_modelling/scripts/1.train_models.py (hunk 14 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/1.train_models.py]: I am down to try it!

File: 1.regression_modelling/scripts/1.train_models.py (hunk 15 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/1.train_models.py]: This is more my personal preference and wanting to make sure that I have each set of features separate, I will keep with even though DNR is happening for now. I think for another PR this can be update, thanks for bringing it up!

**Comment by jenna-tomkinson** [comment]
@MikeLippincott This PR is ready for re-review! Thank you for the first round of comments!

File: 1.regression_modelling/scripts/0.split_data.py (hunk 3 above)
Comment (MikeLippincott) [reply 1.regression_modelling/scripts/0.split_data.py:49]: Question: if removing columns (features) that have NaNs, is this then feature selection? The model is generated on feature selected data and not the normalized data?

File: 1.regression_modelling/scripts/0.split_data.py (hunk 4 above)
Comment (MikeLippincott) [reply 1.regression_modelling/scripts/0.split_data.py:65]: Ahhhh, I see. This makes sense. Documentation would help.

I think a small figure in a future PR showing an example of a non-square "filtered out" crop would be great

File: 1.regression_modelling/scripts/0.split_data.py (hunk 6 above)
Comment (MikeLippincott) [reply 1.regression_modelling/scripts/0.split_data.py]: Amazing!

File: 1.regression_modelling/scripts/0.split_data.py (hunk 7 above)
Comment (MikeLippincott) [reply 1.regression_modelling/scripts/0.split_data.py:134]: Awesome, thank you for clarifying!

File: 1.regression_modelling/scripts/0.split_data.py (hunk 3 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/0.split_data.py:49]: It is generated on the normalized data because we don't perform any variance thresholding or correlation thresholding like we traditionally would on a feature-selected dataset.

We are just removing columns with Nans, not trying to find the least redundant features.

**Review by MikeLippincott — APPROVED** [review]
LGTM! Thank you for addressing comments. Ready to merge!

File: 1.regression_modelling/scripts/0.split_data.py (hunk 3 above)
Comment (MikeLippincott) [reply 1.regression_modelling/scripts/0.split_data.py:49]: Ahhh, gotcha gotcha!

**Comment by jenna-tomkinson** [comment]
@MikeLippincott Thank you for the review! I will be merging now!
